### ISSUE TITLE
[Feature Request] Differentiate actual paragraphs and LUTE's page splitting

### ISSUE THREAD

**Issue by lef-est** [LuteOrg/lute-v3#475]
**Is your feature request related to a problem? Please describe.**

Where to break a parapraph is a stylistic choice by the author and may aid in storytelling. This information is lost because LUTE splits paragraphs to meet the token count per page. 

**Describe the solution you'd like**

Indent original paragraphs, while leaving the "new paragphs" resutled from LUTE's splitting unchanged the same way they are treated now.
If possible, I'd like LUTE to not remove empty lines (or all whitespaces) for the same reason. Many authors use 2 or 3 lines for sub-chapter breaks. Some even differentiate the use of 2 lines or 3 lines.

**Describe alternatives you've considered**

During book creation, give users the option of "don't split paragraphs" (unchecked by default). Not the best solution performance-wise and probably takes more work.

**Additional context**

With LUTE we already lose a lot of information such as bold, italic, underline, images, tables. I understand the difficulty of incorporating them, which entails fundamental changes. But paragraph breaks and empty lines are probably the easiest ones to handle and deserve a chance.

**Comment by jzohrab** [comment]
Lute splits by sentence to keep the token count reasonable: some paragraphs can get _really long_, or the formatting can get weird when people copy/paste text from different places.  It's hard to say what the best solution is here, I still feel that splitting paragraphs up is necessary, and I don't have a great answer for this ... unless I do something like first try to group by paragraphs, and then check the page size and only split the paragraph if the page is, say, 50% longer than the maximum size.  It's a bit convoluted and tricky, but maybe that would suffice.  Does that seem reasonable?

> If possible, I'd like LUTE to not remove empty lines (or all whitespaces) for the same reason.

Yes this sounds reasonable and I'm not sure why I did this in the first place. :-)

> bold, italic, underline, images, tables. 

Yeah that's so tough!  Would be nice though, I agree.  There's an issue to allow for markdown on import of text files (though tables still wouldn't be possible, b/c it's bananas)

Let me know re the paragraph grouping and page size threshold idea ... it could be tricky-ish.

**Comment by LangNerd23** [comment]
Allthough the question was not adressed to me: I personally like this idea with a two step process (First grouping paragraphs and only further split if it exceeds a certain threshold).

At the moment I still have quit a lot of work in pre-processing books, often setting manual separators (---) to avoid parts of the problems mentoined above.

**Comment by jzohrab** [comment]
I've had annoying paragraph splits as well, in places that didn't make sense and complicated the reading of a tough book.

**Comment by jzohrab** [comment]
I've been thinking about this one on and off for a few days, and don't have a _perfect_ solution for it yet, but probably good enough: I'm considering doing grouping based on paragraphs, and just not bothering to split paragraphs at all, even if the paragraph is huge.  For the most part, that should be fine ... if someone is reading James Joyce or Faulkner with Lute, they're probably outside of the target user base anyway.  And such users could just add a new page and split a long paragraph manually if they wanted to.

With this change, the "max words" thing would become a "threshold", and paragraphs would be added to the current page until the threshold is crossed, at which point a new page would be started.  The threshold could be exceeded by any number of tokens -- but it's probably good enough.

**Comment by lef-est** [comment]
> It's hard to say what the best solution is here, I still feel that splitting paragraphs up is necessary, and I don't have a great answer for this ... 

I agree. That's why I put "do not split" under "alternative solution you've considered" :) You probably misunderstood what I presented as the first solution: _"Indent original paragraphs, while leaving the "new paragphs" resutled from LUTE's splitting unchanged the same way they are treated now."_ It means making original paragraphs (indented) and the ones created by LUTE (flush) look different, so users can tell.

> I've been thinking about this one on and off for a few days, and don't have a perfect solution for it yet, but probably good enough: I'm considering doing grouping based on paragraphs, and just not bothering to split paragraphs at all, even if the paragraph is huge. For the most part, that should be fine ... if someone is reading James Joyce or Faulkner with Lute, they're probably outside of the target user base anyway. And such users could just add a new page and split a long paragraph manually if they wanted to.

> With this change, the "max words" thing would become a "threshold", and paragraphs would be added to the current page until the threshold is crossed, at which point a new page would be started.

Yes! I'd love to see this implemented!

> unless I do something like first try to group by paragraphs, and then check the page size and only split the paragraph if the page is, say, 50% longer than the maximum size. It's a bit convoluted and tricky, but maybe that would suffice. Does that seem reasonable?

> Let me know re the paragraph grouping and page size threshold idea ... it could be tricky-ish.

It'll give the best looking pages but as you said it is convoluted. It sounds like a lot to ask, so I proposed the least-effort solution haha. I'd be happy with any improvement.
Sorry for the late reply. I forgot to check notifications.

**Comment by jzohrab** [comment]
Blocked by #562 , "refactor book creation" - when that's done, this will be clearer to work on.

**Comment by jzohrab** [comment]
Lanched in 3.9.2
